File: backend/shared/schemas/school.py
"""Pydantic schemas for School."""

from pydantic import BaseModel, EmailStr, Field, field_validator
from datetime import datetime
from typing import Optional


class SchoolBase(BaseModel):
    """Base schema for School with common fields."""

    name: str = Field(..., min_length=1, max_length=200, description="School name")
    code: str = Field(
        ...,
        min_length=3,
        max_length=50,
        pattern=r"^[A-Za-z0-9-]+$",
        description="Unique school code (letters, numbers, and hyphens only - will be normalized to uppercase)",
    )

    @field_validator("code")
    @classmethod
    def normalize_code(cls, v: str) -> str:
        """Normalize school code to uppercase."""
        return v.upper()
    address: Optional[str] = Field(None, max_length=500, description="School address")
    phone: Optional[str] = Field(
        None,
        pattern=r"^\+?[0-9]{10,15}$",
        description="School phone number",
    )
    email: Optional[EmailStr] = Field(None, description="School email address")


class SchoolCreate(SchoolBase):
    """Schema for creating a new school."""

    pass


class SchoolUpdate(BaseModel):
    """Schema for updating school information (code is immutable)."""

    name: Optional[str] = Field(None, min_length=1, max_length=200)
    address: Optional[str] = Field(None, max_length=500)
    phone: Optional[str] = Field(None, pattern=r"^\+?[0-9]{10,15}$")
    email: Optional[EmailStr] = None


class SchoolResponse(SchoolBase):
    """Schema for school response."""

    id: int
    is_deleted: bool
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


class AdminUserDetails(BaseModel):
    """Schema for admin user details during school registration."""

    email: EmailStr = Field(..., description="Admin email address")
    first_name: str = Field(..., min_length=1, max_length=100, description="Admin first name")
    last_name: str = Field(..., min_length=1, max_length=100, description="Admin last name")
    password: str = Field(
        ...,
        min_length=8,
        max_length=72,
        description="Admin password (must be at least 8 characters and no more than 72 bytes with uppercase, lowercase, digit, and special character)"
    )


class SchoolRegistrationWithAdmin(SchoolBase):
    """Schema for school registration with admin user creation."""

    admin: AdminUserDetails = Field(..., description="Admin user details for the school")


class SchoolRegistrationResponse(SchoolResponse):
    """Schema for school registration response including admin user info."""

    admin_user: dict = Field(..., description="Created admin user information (without password)")

File: frontend/lib/validations/school.ts
/**
 * Zod validation schemas for School entities.
 * 
 * Matches the backend Pydantic schemas in shared/schemas/school.py
 */

import { z } from 'zod';
import {
  adminAccountSchema,
  type AdminAccountFormData,
} from '@/lib/validations/user';

/**
 * School registration form validation schema.
 * 
 * Validation rules:
 * - name: Required, 1-200 characters
 * - code: Required, 3-50 characters, pattern: ^[A-Za-z0-9-]+$
 * - address: Optional, max 500 characters
 * - phone: Optional, pattern: ^\+?[0-9]{10,15}$
 * - email: Optional, valid email format
 */
export const schoolRegistrationSchema = z.object({
  name: z
    .string()
    .min(1, 'School name is required')
    .max(200, 'School name must be less than 200 characters'),
  
  code: z
    .string()
    .min(3, 'School code must be at least 3 characters')
    .max(50, 'School code must be less than 50 characters')
    .regex(
      /^[A-Za-z0-9-]+$/,
      'School code can only contain letters, numbers, and hyphens'
    ),
  
  address: z
    .string()
    .max(500, 'Address must be less than 500 characters')
    .optional()
    .or(z.literal('')),
  
  phone: z
    .string()
    .regex(
      /^\+?[0-9]{10,15}$/,
      'Phone number must be 10-15 digits (optional + prefix)'
    )
    .optional()
    .or(z.literal('')),
  
  email: z
    .string()
    .email('Please enter a valid email address')
    .optional()
    .or(z.literal('')),
});

export type SchoolRegistrationFormData = z.infer<typeof schoolRegistrationSchema>;

/**
 * Combined school registration with admin user schema.
 * This combines school data with admin user data for a single registration flow.
 */
export const schoolRegistrationWithAdminSchema = schoolRegistrationSchema.extend({
  admin: z.object({
    email: z
      .string()
      .min(1, 'Email is required')
      .email('Please enter a valid email address'),
    first_name: z
      .string()
      .min(1, 'First name is required')
      .max(100, 'First name must be less than 100 characters'),
    last_name: z
      .string()
      .min(1, 'Last name is required')
      .max(100, 'Last name must be less than 100 characters'),
    password: z
      .string()
      .min(8, 'Password must be at least 8 characters')
      .max(72, 'Password cannot be longer than 72 bytes')
      .refine(
        (val) => {
          const encoder = new TextEncoder();
          const bytes = encoder.encode(val);
          return bytes.length <= 72;
        },
        {
          message: 'Password cannot be longer than 72 bytes. Please use a shorter password.',
        }
      )
      .regex(/[A-Z]/, 'Password must contain an uppercase letter')
      .regex(/[a-z]/, 'Password must contain a lowercase letter')
      .regex(/[0-9]/, 'Password must contain a number')
      .regex(/[!@#$%^&*()_+\-=\[\]{}|;:,.<>?]/, 'Password must contain a special character (!@#$%^&*)'),
    confirmPassword: z.string().min(1, 'Please confirm your password'),
  }).refine((data) => data.password === data.confirmPassword, {
    message: "Passwords don't match",
    path: ['confirmPassword'],
  }),
});

export type SchoolRegistrationWithAdminFormData = z.infer<typeof schoolRegistrationWithAdminSchema>;

